Rewritten Wolfram Mathematica language parser from scratch

Upabjojr opened the pull request:
New Wolfram Mathematica language parser.

In theory, it should be able to parse everything.

<!-- BEGIN RELEASE NOTES -->
* parsing
  * New powerful Mathematica parser replacing the old one. Language support greatly improved.
<!-- END RELEASE NOTES -->

Upabjojr committed:
Added experimental new Wolfram Mathematica language parser

Upabjojr committed:
fixing types for mypy

Upabjojr committed:
more patterns are parseable

Upabjojr commented:
Any objection to merging this? (All new functionality is hidden).

oscarbenjamin commented:
> Any objection to merging this? (All new functionality is hidden).

If all new functionality is hidden then what is the plan for this in the future?

If this shouldn't be public functionality then another possibility is that it could be in another package.

asmeurer commented:
A little more documentation on this would be nice. Just what sort of parser is this?

Upabjojr commented:
> If all new functionality is hidden then what is the plan for this in the future?

It is a draft for a much more powerful parser for the Wolfram Mathematica language. It would be nice to import Mathematica scripts into SymPy.

By the way, some code from the Rubi module that could be useful to all of SymPy is being save here.

> If this shouldn't be public functionality then another possibility is that it could be in another package.

This is not public yet, as it needs more testing and development. I would suggest to make it the default Mathematica parser once it is more stable. It is much more powerful than the current one.

Upabjojr commented:
> A little more documentation on this would be nice. Just what sort of parser is this?

Not sure, after tokenizing, it matches the brackets first, then proceeds to collapse the operators into AST objects starting from the higher priority ones.

The parser is meant to behave like the `FullForm` function in Mathematica (which outputs the input expression as an AST).

I haven't written any documentation as the code may not yet be final.

oscarbenjamin commented:
> > If all new functionality is hidden then what is the plan for this in the future?
> 
> It is a draft for a much more powerful parser for the Wolfram Mathematica language. It would be nice to import Mathematica scripts into SymPy.

I agree that it would be nice. I don't really understand why a draft implementation would be merged into the sympy codebase though. If it's not going to be public in the codebase then why should it be in the codebase rather than in a separate repo or a gist or even just as an incomplete and unmerged PR?

The normal expectation with a project like sympy that has a large amount of code that needs to be maintained and many downstream users is that something should only be added to the codebase when it is ready.

Upabjojr committed:
support for missing asterisk for Mathematica multiplications

Upabjojr commented:
By the way, I've found some bugs on the current Mathematica parser.

Upabjojr commented:
> The normal expectation with a project like sympy that has a large amount of code that needs to be maintained and many downstream users is that something should only be added to the codebase when it is ready.

OK, let's wait until it's ready then.

Upabjojr committed:
Improving parser for square bracket enclosures, SyntaxError exception support added

Upabjojr committed:
Support for #&[x] expression by allowing precedence inversion of operators.

Operators +- and */ are now mutually associative.

Upabjojr commented on sympy/parsing/tests/test_mathematica.py:
```diff
@@ -64,3 +71,156 @@ def test_mathematica():
+    assert chain("a/b/c*d") == ["Times", "a", ["Power", "b", "-1"], ["Power", "c", "-1"], "d"]
+    assert chain("a/b/c") == ["Times", "a", ["Power", "b", "-1"], ["Power", "c", "-1"]]
+    assert chain("a-b-c") == ["Plus", "a", ["Times", "-1", "b"], ["Times", "-1", "c"]]
+    assert chain("1/a") == ["Times", "1", ["Power", "a", "-1"]]
```
@certik @oscarbenjamin do you have access to Wolfram Mathematica?

Can you please check if the output of `FullForm[Hold[1/a]]` is `Times[1, Power[a, -1]]` or `Power[a, -1]`?

certik replied:
I don't have access to Mathematica.

oscarbenjamin replied:
Me neither

Upabjojr commented:
@quasi-human maybe you are interested in this PR.

asmeurer commented:
Have you tried replacing the current parser with this and seeing if all the tests pass? 

I don't really see why this would need to be marked as experimental or anything like that. The only public entry-point to the parser should be the top-level `mathematica` function. The details of the parser can all be internal. I guess we might explicitly document that, but in general, for any parser that we include in SymPy I think we should always keep the parser specifics as internal implementation details. That way we can always switch to a different kind of parser or a better parser library if we want to without having to worry about compatibility breaks. 

The only exception to this is if we make a parser that is designed to be user extensible. I guess the Mathematica parser has some degree of that with the dictionaries like `CORRESPONDENCES`, but there is also `additional_translations` which is what users should be using instead. Is there any reason users might be modifying those dictionaries in the current parser instead of using that keyword argument?

Unless the reason you want it experimental is that you're still unsure about the correctness. But in that case, I would say we should try to do more testing here before merging. Surly there is some way to get a list of test expressions that we can pass to this.

asmeurer commented:
> Have you tried replacing the current parser with this and seeing if all the tests pass?

Actually the [current tests](https://github.com/sympy/sympy/blob/master/sympy/parsing/tests/test_mathematica.py) are pretty sparse. So I would definitely try to expand the test suite here if possible. 

(an aside, a test that mathematica printer -> mathematica parser roundtrips would be a great use-case for [hypothesis testing](https://github.com/sympy/sympy/issues/20914))

Upabjojr commented:
> The only public entry-point to the parser should be the top-level `mathematica` function. The details of the parser can all be internal. I guess we might explicitly document that, but in general, for any parser that we include in SymPy I think we should always keep the parser specifics as internal implementation details. That way we can always switch to a different kind of parser or a better parser library if we want to without having to worry about compatibility breaks.

In reality Wolfram Mathematica has a standardized _FullForm_ of its expressions that is an easy representation of its expressions. Once you have the _FullForm_, it's extremely easy to translate the code into SymPy. I would go as far as suggesting that the _FullForm_ parser could be made public (and not just internal).

The tests I've added are all verifying that the expression is converted into a Python-list form corresponding to Mathematica's _FullForm_.

So far all the code I've added is not publicly documented and therefore hidden from end-users. We have to understand if we want to dare replacing the current parser completely (this new one is supposed to be much more powerful once it's finished).

Upabjojr committed:
New parser is now the default one.

Upabjojr commented:
Deprecation of `additional_translations` https://github.com/sympy/sympy/issues/23042.

Upabjojr committed:
Support parsing functions and some more Mathematica nodes. Commented Mathematica code is now parsed correctly.

Upabjojr commented:
I would say... this PR looks finished now. The old parser has been deprecated and a new much more powerful one has been created.

I believe there should be no more objections to merging this. Are there?

Upabjojr commented on sympy/parsing/mathematica.py:
```diff
@@ -1,28 +1,101 @@
+    if additional_translations is not None:
+        SymPyDeprecationWarning(
+            feature="additional_translations parameter for the Mathematica parser",
+            last_supported_version="1.9",
```
No idea. @oscarbenjamin ?

asmeurer replied:
Why are we removing additional_translations? This is a regression. Without this, it is impossible for people to define custom translations for their own functions. Is it not possible to keep support for this?

Upabjojr replied:
> Without this, it is impossible for people to define custom translations for their own functions. Is it not possible to keep support for this?

New way to do it (much more simple and functional):
`mathematica(" ... ").subs(additional_translation)`

oscarbenjamin replied:
If this is to be deprecated then it should say `last_supported_version="1.10"`.

asmeurer replied:
@oscarbenjamin isn't the release branch already made? If this is merged into master it will go into 1.11. At least release.py in `master` says 1.11.dev. In general the last_supported_version should be whatever is in release.py, minus `.dev`.

oscarbenjamin replied:
Yes, so if this is merged then the feature will be deprecated in 1.11 and then the last supported version will be 1.10.

asmeurer replied:
Hmm, that's not how I've been interpreting this flag, and it's not how I documented it in https://github.com/sympy/sympy/pull/22900. I wonder if others have been doing both. I guess the flag should be renamed to `deprecated_since_version` to make this clearer.

asmeurer replied:
Oh I see. There is `deprecated_since_version` *and* `last_supported_version`. So yeah, this is something I already fixed in https://github.com/sympy/sympy/pull/22900. I removed `last_supported_version` and there's now just `deprecated_since_version`. `last_supported_version` is confusing and completely redundant, and exactly 0 deprecations in master are using it.

So in short, replace this with `deprecated_since_version="1.11"`.

If you want to wait until https://github.com/sympy/sympy/pull/22900 is merged and do this deprecation in the new style, that would be great too. Otherwise if this is merged first, I'll update it in that PR.

asmeurer replied:
By the way, "last_supported_version" was actually supposed to be the last version that would support the feature, i.e., the future version. But we don't even use a version-based deprecation policy, which is why I removed it.

Upabjojr commented on sympy/parsing/mathematica.py:
```diff
@@ -1,28 +1,101 @@
+    >>> e
+    F(7, 5, 3)
+    >>> from sympy import Function, Max, Min
+    >>> e.replace(Function("F"), lambda *x: Max(*x)*Min(*x))
```
@asmeurer this is the new way to deal with additional translations.

Upabjojr committed:
Removed regular expressions to add missing * operator. Better support for CompoundExpression.

Upabjojr committed:
Added support for newlines to the Mathematica parser

Upabjojr committed:
code refactory

Upabjojr commented:
OK, this PR is ready now. Any last reviews?

Upabjojr commented:
@certik @oscarbenjamin concerning the rule-based integration (RUBI) module. If this PR gets merged, we should probably be able to parse the RUBI Mathematica files directly without the need of using Wolfram Mathematica.

Maybe in the future we can get rid of the code generator to translate RUBI into Python. I think this parser combined with the utility functions (i.e. missing functions in SymPy used by RUBI), plus the MatchPy connector could be used to create a pickled package to run RUBI in Python.

@oscarbenjamin at this point maybe we won't need an independent repository for the generated RUBI code.

oscarbenjamin commented:
I would rather not include a pickled package directly in the SymPy release files. I'm not sure if that would be problematic for some downstream distributors.

Is it enough just to parse the wolfram RUBI rules? Surely there also needs to be something that can replicate the semantics of those rules. I don't think matchpy does that and I don't see anything else in SymPy that does.

> @oscarbenjamin at this point maybe we won't need an independent repository for the generated RUBI code.

The primary reason I would want RUBI to be in another repo is that I think it should only be part of the main SymPy codebase when it is actually usable. The changes here don't immediately affect that as far as I can tell.

Upabjojr commented:
> I would rather not include a pickled package directly in the SymPy release files. I'm not sure if that would be problematic for some downstream distributors.

I mean an independent package on PyPI containing just the pickled rule matcher (supposing this can be done).

> Is it enough just to parse the wolfram RUBI rules? Surely there also needs to be something that can replicate the semantics of those rules. I don't think matchpy does that and I don't see anything else in SymPy that does.

No, it is still not enough. Some code of the `sympy.integrals.rubi` module should still be extracted, namely the one regarding the transformation of Mathematica's abstract syntax tree into RUBI rules. In particular:

- `SetDelayed` nodes should become `ReplacementRule` in MatchPy's `ManyToOneReplacer` object.
- `Pattern` nodes should become `WildcardDot` objects (i.e. SymPy's connection wildcard to MatchPy).
- `Condition` nodes should become `CustomConstraint` objects in MatchPy.

> The changes here don't immediately affect that as far as I can tell.

They are a first step towards solving the issue.

Upabjojr commented:
> Please hold off merging this or any other PR that adds new deprecations until #22900 is merged.

OK!

Upabjojr committed:
Merge branch 'master' into mathematica_parser_full_form

Upabjojr committed:
Adapt to new deprecation policy

Upabjojr commented on doc/src/explanation/active-deprecations.md:
````diff
@@ -74,6 +74,37 @@ will need to either add a `warnings` filter as above or use pytest to filter
+>>> from sympy import Function, Max, Min
+>>> mathematica("F[7,5,3]").replace(Function("F"), lambda *x: Max(*x)*Min(*x))
+21
+```
````
@asmeurer did I follow the new deprecation policy correctly?

asmeurer replied:
Yes, it looks good. Thanks.
